sys/kern/vm_amap.c

```diff
@@ -153,14 +196,60 @@ void vm_amap_remove_pages(vm_aref_t aref, size_t start, size_t nslots) {
 }
 
 void vm_amap_hold(vm_amap_t *amap) {
-  refcnt_acquire(&amap->ref_cnt);
+  SCOPED_MTX_LOCK(&amap->mtx);
+  amap->ref_cnt++;
 }
 
 void vm_amap_drop(vm_amap_t *amap) {
-  if (refcnt_release(&amap->ref_cnt)) {
-    vm_amap_remove_pages_unlocked(amap, 0, amap->slots);
-    kfree(M_AMAP, amap->pg_list);
-    kfree(M_AMAP, amap->pg_bitmap);
-    pool_free(P_VM_AMAP_STRUCT, amap);
+  WITH_MTX_LOCK (&amap->mtx) {
+    amap->ref_cnt--;
+    if (amap->ref_cnt >= 1)
+      return;
+  }
+  vm_amap_remove_pages_unlocked(amap, 0, amap->slots);
+  kfree(M_AMAP, amap->anon_list);
+  kfree(M_AMAP, amap->anon_bitmap);
+  pool_free(P_VM_AMAP_STRUCT, amap);
+}
+
+static vm_anon_t *alloc_empty_anon(void) {
+  vm_anon_t *anon = pool_alloc(P_VM_ANON_STRUCT, M_WAITOK);
+  anon->ref_cnt = 1;
+  anon->page = NULL;
+  mtx_init(&anon->mtx, MTX_SLEEP);
+  return anon;
+}
+
+vm_anon_t *vm_anon_alloc(void) {
+  vm_anon_t *anon = alloc_empty_anon();
+  vm_page_t *page = vm_page_alloc(1);
+  if (!page) {
+    vm_anon_drop(anon);
+    return NULL;
   }
+  pmap_zero_page(page);
+  anon->page = page;
+  return anon;
+}
+
+void vm_anon_hold(vm_anon_t *anon) {
+  SCOPED_MTX_LOCK(&anon->mtx);
+  anon->ref_cnt++;
+}
+
+void vm_anon_drop(vm_anon_t *anon) {
+  WITH_MTX_LOCK (&anon->mtx) {
+    anon->ref_cnt--;
+    if (anon->ref_cnt >= 1)
+      return;
+  }
+  vm_page_free(anon->page);
+  pool_free(P_VM_ANON_STRUCT, anon);
+}
```

Comment:
Perfect places for atomics :)

Reply:
I remember that there was some reason why I changed it for atomics. I will think about it once again, because maybe that problem was fixed (I have made few changes after that change). I will change it to atomics if it is possible.